File: Python/自己搞定/Medium/739. Daily Temperatures.py
# 這題反而很簡單，去找下一個比自己高的值
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        stack=[] ;ans = [0]*len(temperatures)
        # 一樣做Monotonic Stack，但這邊不一樣直接把找到的大值減去stack的index
        for index in range(len(temperatures)):
            while (stack) and (temperatures[stack[-1]]<temperatures[index]):
                # 當遇到比原本的大取出來，並且拿來跟現在的index比算距離
                popindex = stack.pop()
                ans[popindex] = index-popindex
            stack.append(index)
        return ans 

File: C++/自己搞定/Medium/739. Daily Temperatures.cpp
#include<iostream>
#include <bits/stdc++.h>
#include <map>
using namespace std;
//要找到哪幾天後的溫度比現在高
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        vector<int> answer(temperatures.size());
        for (int index = temperatures.size() - 1; index >= 0; --index) {
            int nextIndex = index+1;
            while (nextIndex < temperatures.size() && temperatures[nextIndex] <= temperatures[index]) {
                if (answer[nextIndex] > 0)
                    nextIndex = answer[nextIndex] + nextIndex;
                else 
                    nextIndex = temperatures.size();
            }
            if (nextIndex < temperatures.size()) 
                answer[index] = nextIndex - index;
        }
        return answer;
    }
};


class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        std::stack<int> tempStack;std::vector<int> answer(temperatures.size());int nowIndex=0;
        for(int index = 0;index<temperatures.size();index++){
            nowIndex = index-1;
            while(!tempStack.empty()&&temperatures[index]>tempStack.top()){
                if(answer[nowIndex]==0){
                    answer[nowIndex] = index-nowIndex;
                    tempStack.pop();
                }
                --nowIndex;
            }
            tempStack.emplace(temperatures[index]);
        }
        return answer;      
    }
};

